connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -0,0 +1,535 @@
+                .filter(wl -> wl.size() > 0)
+                .collect(Collectors.toList());
+        int existingWorkersNum = existingWorkers.size();
+        int totalWorkersNum = completeWorkerAssignment.size();
+        int newWorkersNum = totalWorkersNum - existingWorkersNum;
+
+        Map<String, ConnectorsAndTasks> revoking = new HashMap<>();
+        if (!(newWorkersNum > 0 && existingWorkersNum > 0)) {
+            log.debug("Business as usual existing {} new {} total {}", existingWorkersNum,
+                    newWorkersNum, totalWorkersNum
+            );
+
+            if (log.isDebugEnabled()) {
+                completeWorkerAssignment.forEach(wl -> log.debug(
+                        "Current plan worker {} connectors {} tasks {}",
+                        wl.worker(),
+                        wl.connectorsSize(),
+                        wl.tasksSize()));
+            }
+            return revoking;
+        }
+
+        log.debug("Free workers {} loaded workers {} total workers {}",
+                newWorkersNum, existingWorkersNum, totalWorkersNum);
+
+        int floorConnectors = totalActiveConnectorsNum / totalWorkersNum;
+        int floorTasks = totalActiveTasksNum / totalWorkersNum;
+
+        log.debug("Old connectors per worker {}", totalActiveConnectorsNum / existingWorkersNum);
+        log.debug("New connectors per worker {}", totalActiveConnectorsNum / totalWorkersNum);
+        log.debug("Old tasks per worker {}", totalActiveTasksNum / existingWorkersNum);
+        log.debug("New tasks per worker {}", totalActiveTasksNum / totalWorkersNum);
+
+        int numToRevoke = floorConnectors;
+        for (WorkerLoad existing : existingWorkers) {
+            Iterator<String> connectors = existing.connectors().iterator();
+            for (int i = existing.connectorsSize(); i > floorConnectors && numToRevoke > 0; --i, --numToRevoke) {
+                ConnectorsAndTasks resources = revoking.computeIfAbsent(
+                    existing.worker(),
+                    w -> ConnectorsAndTasks.embed(new ArrayList<>(), new ArrayList<>()));
+                resources.connectors().add(connectors.next());
+            }
+            if (numToRevoke == 0) {
+                break;
+            }
+        }
+
+        numToRevoke = floorTasks;
+        for (WorkerLoad existing : existingWorkers) {
+            Iterator<ConnectorTaskId> tasks = existing.tasks().iterator();
+            for (int i = existing.tasksSize(); i > floorTasks && numToRevoke > 0; --i, --numToRevoke) {
+                ConnectorsAndTasks resources = revoking.computeIfAbsent(
+                    existing.worker(),
+                    w -> ConnectorsAndTasks.embed(new ArrayList<>(), new ArrayList<>()));
+                resources.tasks().add(tasks.next());
+            }
+            if (numToRevoke == 0) {
+                break;
+            }
+        }
+
+        // TODO: improve logging messages in this class
```

Comment:
+1. The value of logging in this whole class is not really going to be knowing the final end state, but being able to track the logic in this class to figure out why the end state doesn't match an expected state on some user's Connect cluster, and hopefully being able to learn enough to understand what paths are taken so that we can reproduce the case.

The other approach is to log all of the initial state up front and all of the new state at the end, such that with that information we could reproduce any logged scenario with a unit test that we can then debug to trace the logic.

Reply:
I've added and improved debug logs. If during additional testing we consider that some info logs are warranted, we could return here.